### DISCUSSION TITLE
Lambda Function parameter infer type - External Struct

### DISCUSSION THREAD

**Discussion by Alexzander-Stone** [aptos-labs/aptos-developer-discussions#290] [Questions]
### Discord user ID

spielcrypto.apt

### Describe your question in detail.

My `Token` Struct parameter can't have it's `id` field accessed.
`vector::for_each`'s lambda parameter's type can not be inferred.

### What error, if any, are you getting?
```rust
error[E04010]: cannot infer type
   ┌─ /home/spielcrypto/ghost-lost/ghost-lost/ghostly-contract/sources/ghostly.move:17:69
   │
16 │         vector::for_each<Token>(nfts, |nft| {
   │                                        --- Could not infer the type before field access. Try annotating here
17 │             vector::push_back<TokenId>(&mut trackUserAssets.tokens, nft.id);
   │                                                                     ^^^^^^ Unbound field 'id'

{
  "Error": "Move compilation failed: Compilation error"
}
```

### Which operating system are you using?

N/A

### Which SDK or tool are you using? (if any)

Aptos CLI

**Comment by Alexzander-Stone** [comment]
# Inferring type from a lambda expression's Struct parameter
Hey `spielcrypto.apt`, when accessing a Struct's fields not owned by your module, you'll need to use available accessors on the Struct's originating module.
For `Token`, you can make use of [`get_token_id`](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token/sources/token.move#L1344-L1347).

# Applied to your use-case
```rust
vector::for_each<Token>(nfts, |nft| {
    let id = token::get_token_id(&nft);
    vector::push_back<TokenId>(&mut trackUserAssets.tokens, id);
    // ......
}
```

**Comment by spielcrypto** [comment]
Yes that works thanks a lot!
The only thing is that the error ` Could not infer the type before field access. Try annotating here` is misleading

**Comment by spielcrypto** [comment]
but now if I have this:
```rust
vector::for_each<NFT>(nfts, |nft| {
            let token_addr = object::object_address(&nft.token);
            let token = borrow_global<Token>(token_addr);
            let id = get_token_id(token);

            vector::push_back<TokenId>(&mut trackUserAssets.tokens, id);
            deposit_token(account, nft);
        });
```

I get this error:
```rust
error[E04007]: incompatible types
    ┌─ /home/spielcrypto/ghost-lost/ghost-lost/ghostly-contract/sources/ghostly.move:25:9
    │  
 25 │ ╭         vector::for_each<NFT>(nfts, |nft| {
    │                            --- Given: '(ghostly=0x1)::keep::NFT'
 26 │ │             let token_addr = object::object_address(&nft.token);
 27 │ │             let token = borrow_global<Token>(token_addr);
 28 │ │             let id = get_token_id(token);
    · │
 31 │ │             deposit_token(account, nft);
 32 │ │         });
    │ ╰──────────^ Invalid call of '(std=0x1)::vector::for_each'. Invalid argument for parameter 'f'
    │  
    ┌─ /home/spielcrypto/.move/https___github_com_aptos-labs_aptos-core_git_testnet/aptos-move/framework/aptos-token/sources/token.move:955:55
    │
955 │     public fun deposit_token(account: &signer, token: Token) acquires TokenStore {
    │                                                       ----- Expected: '(aptos_token=0x3)::token::Token'

error[E04010]: cannot infer type
   ┌─ /home/spielcrypto/ghost-lost/ghost-lost/ghostly-contract/sources/ghostly.move:26:30
   │
26 │             let token_addr = object::object_address(&nft.token);
   │                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Cannot infer the type parameter `T` for generic function `(aptos_framework=0x1)::object::object_address`. Try providing a type parameter.

error[E04010]: cannot infer type
   ┌─ /home/spielcrypto/ghost-lost/ghost-lost/ghostly-contract/sources/ghostly.move:26:54
   │
25 │         vector::for_each<NFT>(nfts, |nft| {
   │                                      --- Could not infer the type before field access. Try annotating here
26 │             let token_addr = object::object_address(&nft.token);
   │                                                      ^^^^^^^^^ Unbound field 'token'

error[E05001]: ability constraint not satisfied
    ┌─ /home/spielcrypto/ghost-lost/ghost-lost/ghostly-contract/sources/ghostly.move:27:25
    │
 27 │             let token = borrow_global<Token>(token_addr);
    │                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    │                         │             │
    │                         │             The type '(aptos_token=0x3)::token::Token' does not have the ability 'key'
    │                         Invalid call of 'borrow_global'
    │
    ┌─ /home/spielcrypto/.move/https___github_com_aptos-labs_aptos-core_git_testnet/aptos-move/framework/aptos-token/sources/token.move:158:12
    │
158 │     struct Token has store {
    │            ----- To satisfy the constraint, the 'key' ability would need to be added here

error[E04001]: restricted visibility
   ┌─ /home/spielcrypto/ghost-lost/ghost-lost/ghostly-contract/sources/ghostly.move:27:25
   │
27 │             let token = borrow_global<Token>(token_addr);
   │                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   │                         │             │
   │                         │             The type '(aptos_token=0x3)::token::Token' was not declared in the current module. Global storage access is internal to the module'
   │                         Invalid call to borrow_global.

```
The NFT struct is this:

```rust
    struct NFT has key, copy, drop {
        token: Object<Token>,
        collection: Object<Collection>,
    }
```
It is like inside for_each I cannot get any prop from any `Struct`

**Reply by Alexzander-Stone** [reply]
# Overall
I'd recommend reading through the resources at the end to see if the DA standard fits your needs, before committing to the legacy Token standard.

---

```rust
vector::for_each<NFT>(nfts, |nft| {
            let token_addr = object::object_address(&nft.token);
            let token = borrow_global<Token>(token_addr);
            let id = get_token_id(token);

            vector::push_back<TokenId>(&mut trackUserAssets.tokens, id);
            deposit_token(account, nft);
});
```
Problems:
1. `borrow_global` can't be used against a Struct that is defined outside your module. In this instance, `borrow_global<Token>` contains `Token` which is defined in the Aptos-Token module.
2. `object::object_address(&nft.token)` will need a type parameter defined.
3. Your `deposit_token(account: &signer, token: Token)` has the 2nd parameter defined as a `Token`, but you're attempting to pass it an `NFT`.

# Legacy Token Solution
Instead of storing the `Object<Token>` in your Struct, store just the `Token`. Then, you can call `get_token_id`, without needing to use `borrow_global`.

```rust
struct NFT has key, copy, drop {
    token: Token,
    collection: Object<Collection>,
}
```
```rust
vector::for_each<NFT>(nfts, |nft| {
            let NFT { token, collection } = nft;
            let id = get_token_id(&token);

            vector::push_back<TokenId>(&mut trackUserAssets.tokens, id);
            deposit_token(account, token); 
});
```

If there was a pre-planned reason for storing as an `Object<Token>`,  let me know!

# Digital Asset standard solution
Since Aptos Token is a legacy standard, it might be worth it for you to use the Digital Asset standard, instead.
Digital Assets have an available accessor to [borrow](https://legacy.aptos.dev/reference/move/?branch=mainnet&page=aptos-token-objects/doc/token.md#0x4_token_borrow:~:text=Implementation-,Function%20borrow,-fun%20borrow%3CT) the `Token` from within you module.

It'll look something similar to:
```rust
vector::for_each<NFT>(nfts, |nft| {
            let NFT { token, collection } = nft;
            let id =  borrow<Token>(&token).id;

            vector::push_back<TokenId>(&mut trackUserAssets.tokens, id);
            deposit_token(account, token); 
});
```



# Additional Resources
* [Digital Asset Aptos standard](https://aptos.dev/en/build/smart-contracts/aptos-standards#digital-asset-da)
* [Aptos Learn Digital Asset overview](https://learn.aptoslabs.com/example/aptogotchi-intermediate/digital-assets)
* [Digital Asset source code](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-token-objects/sources/token.move)

**Reply by spielcrypto** [reply]
In then end following your suggestion of using [Digital Asset Aptos standard](https://aptos.dev/en/build/smart-contracts/aptos-standards#digital-asset-da) I did this:

```rust
public fun send_to_ghostly_world(account: &signer, nfts: vector<Object<Token>>) acquires TrackUserAssets {
        let account_addr = signer::address_of(account);
        let trackUserAssets = borrow_global_mut<TrackUserAssets>(account_addr);

        vector::for_each<Object<Token>>(nfts, |nft| {
            let token_addr = object::object_address(&nft);

            vector::push_back<address>(&mut trackUserAssets.tokens, token_addr);
            object::transfer(account, nft, @ghostly);
        });
    }
```

but now I want to implement the claim feature where contract send back the digital asset. If the method is this one
` object::transfer(signer, nft, to);` how do I represent the contract as a signer?

**Reply by spielcrypto** [reply]
and now I change to this:
```rust
public fun send_to_ghostly_world(account: &signer, nfts: vector<Object<Token>>) {
        let account_addr = signer::address_of(account);

        let constructor_ref = object::create_object(account_addr);


        let token_signer = object::generate_signer(&constructor_ref);
        
        vector::for_each<Object<Token>>(nfts, |nft| {
            let track_user_asset = TrackUserAsset{
                token: option::some(nft)
            };

            move_to(account, track_user_asset);

            let track_user_asset_obj = object::address_to_object<TrackUserAsset>(signer::address_of(&token_signer));

            object::transfer_to_object(account, nft, track_user_asset_obj);
        });
    }
```

In this case can I recover all the objects created by the signer?

**Reply by Alexzander-Stone** [reply]
Hi @dancespiele, it might help if you describe what you're attempting to solve with each method in your module. Right now for me, it's a bit confusing why you're repeatedly calling `move_to` within your `vector::for_each` call.

## Digital Asset examples

There's a good deal of examples within the [primary `aptos-core` repo](https://github.com/aptos-labs/aptos-core/tree/main/aptos-move/move-examples). Here's a few:
* [Token object examples](https://github.com/aptos-labs/aptos-core/tree/b1b6b050496eb5b7ff09ae18524df2f23b6fd2cf/aptos-move/move-examples/token_objects)
* [Marketplace](https://github.com/aptos-labs/aptos-core/tree/b1b6b050496eb5b7ff09ae18524df2f23b6fd2cf/aptos-move/move-examples/marketplace)

Along with the Aptos Learn examples I linked, prior:
* [Aptogotchi Intermediate: Token Standards - Digital Assets](https://learn.aptoslabs.com/example/aptogotchi-intermediate/digital-assets) 
* [NFT Marketplace](https://learn.aptoslabs.com/example/nft-marketplace/demo)

**Reply by spielcrypto** [reply]
Yes sure, I want something simple, a contract which have 2 entry functions, one where user can send NFTs to the contract and other to recover the NFTs that the user sent

**Reply by Alexzander-Stone** [reply]
If that's the case @dancespiele, then I think using the [NFT Marketplace](https://learn.aptoslabs.com/example/nft-marketplace/demo) as your base template and making modifications to it is your best bet!

At the very least, understanding how [Listings](https://learn.aptoslabs.com/example/nft-marketplace/object-resources#:~:text=Resource-,Listing,-object%3A%20The%20token) work (and what needs to be modified for just holding the NFT) and the [entry functions](https://learn.aptoslabs.com/example/nft-marketplace/entry-functions) will get you to 90% of the way there.
